### PR TITLE
114 root has section most popular items

### LINKED ISSUE #114: Extension: Viewing most popular items
on /
see listing of most popular items in a section above the main items directory

### PR DESCRIPTION
**Pull request by meyerhoferc**
* adds feature: most popular items on items index
* It finds the most popular items by searching through the order-items table and orders by the count of the item id, which represents the total number of times that item has been ordered (not dependent on the quantity in the order)
* I limited the most popular items to six on the root path, this can be easily changed in the Item model

* Do we want items to duplicate? That is, should an item that's in the most popular category also be listed in the items section on the root path?

closes #114

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@ class ItemsController < ApplicationController
 
   def index
     @items = Item.all
+    @popular_items = Item.most_popular
```
Comment (cmacaulay) [inline app/controllers/items_controller.rb:5]: Love that this is being passed through in the controller.

File: app/models/item.rb [hunk 2]
```diff
@@ -24,6 +24,13 @@ def get_order_item(order)
   def get_subtotal_on_order(order)
     get_order_item(order).subtotal
   end
+
+  def self.most_popular
+    item_ids = select('items.id, count(order_items.item_id) as frequency')
+               .joins(:order_items).group('items.id')
+               .order('frequency desc').limit(6)
+    item_ids.map { |item| find(item.id) }
+  end
```
Comment (cmacaulay) [inline app/models/item.rb:33]: 🎉

File: app/views/items/index.html.erb [hunk 3]
```diff
@@ -1 +1,30 @@
...
+            <% else %>
+              <button  class="btn disabled">Retired</button>
+            <% end %>
+        </div>
+      </div>
+    </div>
+  <% end %>
+</div>
+<div class="divider"></div>
 <%= render partial: 'shared/items' %>
```
Comment (cmacaulay) [inline app/views/items/index.html.erb:30]: 👍

File: spec/features/items/user_can_view_all_items_spec.rb [hunk 4]
```diff
@@ -2,14 +2,60 @@
...
+    OrderItem.create!(order: order2,
+                      item: item3,
+                      quantity: 4,
+                      price: item3.price)
+    OrderItem.create!(order: order2,
+                      item: item4,
+                      quantity: 1,
+                      price: item4.price)
+    visit root_path
+
```
Comment (cmacaulay) [inline spec/features/items/user_can_view_all_items_spec.rb:45]: @meyerhoferc @valerietrudell @CjMoore thoughts on adding an OrderItem fabricator? Happy to take that on to get some more experience in fabricator.

**Review by cmacaulay — COMMENTED** [review]
Looks good! I think we should duplicate it, and maybe add a link/button to the top of the items index "View most popular items". Thoughts?

File: spec/features/items/user_can_view_all_items_spec.rb (hunk 4 above)
Comment (meyerhoferc) [reply spec/features/items/user_can_view_all_items_spec.rb:45]: there is one but it doesn't work. it would definitely make our tests more clean

Comment (CjMoore) [reply spec/features/items/user_can_view_all_items_spec.rb:45]: I agree. I'll play around with it and try to make it work.

File: spec/features/items/user_can_view_all_items_spec.rb [hunk 5]
```diff
@@ -2,14 +2,60 @@
 
 describe "When user visits root" do
   scenario "They see list of all items" do
-    item = Fabricate(:item)
-    item2 = Fabricate(:item)
-
+    item, item2 = Fabricate.times(2, :item)
```
Comment (CjMoore) [inline spec/features/items/user_can_view_all_items_spec.rb:5]: 👍

File: spec/features/orders/user_can_view_all_of_their_orders_spec.rb [hunk 6]
```diff
@@ -5,15 +5,23 @@
...
+                      quantity: 1,
+                      price: item1.price)
+    OrderItem.create!(order: order1,
+                      item: item2,
+                      quantity: 1,
+                      price: item2.price)
+    OrderItem.create!(order: order2,
+                      item: item1,
+                      quantity: 2,
+                      price: item1.price)
```
Comment (CjMoore) [inline spec/features/orders/user_can_view_all_of_their_orders_spec.rb:24]: 👍

**Review by CjMoore — APPROVED** [review]
Looks great.

**Pull request merged**